File: clients/ruby/model/action_order.rb
module Model

# Order to perform an action for unit
class ActionOrder
    # Read ActionOrder from input stream
    def self.read_from(stream)
        tag = stream.read_int()
        if tag == ActionOrder::Pickup::TAG
            return ActionOrder::Pickup.read_from(stream)
        end
        if tag == ActionOrder::UseShieldPotion::TAG
            return ActionOrder::UseShieldPotion.read_from(stream)
        end
        if tag == ActionOrder::DropShieldPotions::TAG
            return ActionOrder::DropShieldPotions.read_from(stream)
        end
        if tag == ActionOrder::DropWeapon::TAG
            return ActionOrder::DropWeapon.read_from(stream)
        end
        if tag == ActionOrder::DropAmmo::TAG
            return ActionOrder::DropAmmo.read_from(stream)
        end
        if tag == ActionOrder::Aim::TAG
            return ActionOrder::Aim.read_from(stream)
        end
        raise "Unexpected tag value"
    end

    # Pick up loot
    class Pickup
        TAG = 0
    
        # Loot id
        attr_accessor :loot
    
        def initialize(loot)
            @loot = loot
        end
    
        # Read Pickup from input stream
        def self.read_from(stream)
            loot = stream.read_int()
            Pickup.new(loot)
        end
    
        # Write Pickup to output stream
        def write_to(stream)
            stream.write_int(TAG)
            stream.write_int(@loot)
        end
    
        def to_s
            string_result = "Pickup { "
            string_result += "loot: "
            string_result += @loot.to_s
            string_result += " }"
            string_result
        end
    
        def to_str
            to_s
        end
    end
    # Use shield potion
    class UseShieldPotion
        TAG = 1
    
    
        def initialize()
        end
    
        # Read UseShieldPotion from input stream
        def self.read_from(stream)
            UseShieldPotion.new()
        end
    
        # Write UseShieldPotion to output stream
        def write_to(stream)
            stream.write_int(TAG)
        end
    
        def to_s
            string_result = "UseShieldPotion { "
            string_result += " }"
            string_result
        end
    
        def to_str
            to_s
        end
    end
    # Drop shield potions on the ground
    class DropShieldPotions
        TAG = 2
    
        # Amount of potions
        attr_accessor :amount
    
        def initialize(amount)
            @amount = amount
        end
    
        # Read DropShieldPotions from input stream
        def self.read_from(stream)
            amount = stream.read_int()
            DropShieldPotions.new(amount)
        end
    
        # Write DropShieldPotions to output stream
        def write_to(stream)
            stream.write_int(TAG)
            stream.write_int(@amount)
        end
    
        def to_s
            string_result = "DropShieldPotions { "
            string_result += "amount: "
            string_result += @amount.to_s
            string_result += " }"
            string_result
        end
    
        def to_str
            to_s
        end
    end
    # Drop current weapon
    class DropWeapon
        TAG = 3
    
    
        def initialize()
        end
    
        # Read DropWeapon from input stream
        def self.read_from(stream)
            DropWeapon.new()
        end
    
        # Write DropWeapon to output stream
        def write_to(stream)
            stream.write_int(TAG)
        end
    
        def to_s
            string_result = "DropWeapon { "
            string_result += " }"
            string_result
        end
    
        def to_str
            to_s
        end
    end
    # Drop ammo
    class DropAmmo
        TAG = 4
    
        # Weapon type index (starting with 0)
        attr_accessor :weapon_type_index
        # Amount of ammo
        attr_accessor :amount
    
        def initialize(weapon_type_index, amount)
            @weapon_type_index = weapon_type_index
            @amount = amount
        end
    
        # Read DropAmmo from input stream
        def self.read_from(stream)
            weapon_type_index = stream.read_int()
            amount = stream.read_int()
            DropAmmo.new(weapon_type_index, amount)
        end
    
        # Write DropAmmo to output stream
        def write_to(stream)
            stream.write_int(TAG)
            stream.write_int(@weapon_type_index)
            stream.write_int(@amount)
        end
    
        def to_s
            string_result = "DropAmmo { "
            string_result += "weapon_type_index: "
            string_result += @weapon_type_index.to_s
            string_result += ", "
            string_result += "amount: "
            string_result += @amount.to_s
            string_result += " }"
            string_result
        end
    
        def to_str
            to_s
        end
    end
    # Start/continue aiming
    class Aim
        TAG = 5
    
        # Shoot (only possible in full aim)
        attr_accessor :shoot
    
        def initialize(shoot)
            @shoot = shoot
        end
    
        # Read Aim from input stream
        def self.read_from(stream)
            shoot = stream.read_bool()
            Aim.new(shoot)
        end
    
        # Write Aim to output stream
        def write_to(stream)
            stream.write_int(TAG)
            stream.write_bool(@shoot)
        end
    
        def to_s
            string_result = "Aim { "
            string_result += "shoot: "
            string_result += @shoot.to_s
            string_result += " }"
            string_result
        end
    
        def to_str
            to_s
        end
    end
end

end

File: clients/python/model/action_order.py
from stream_wrapper import StreamWrapper

class ActionOrder:
    """Order to perform an action for unit"""

    @staticmethod
    def read_from(stream: StreamWrapper) -> "ActionOrder":
        """Read ActionOrder from input stream
        """
        tag = stream.read_int()
        if tag == Pickup.TAG:
            return ActionOrder.Pickup.read_from(stream)
        if tag == UseShieldPotion.TAG:
            return ActionOrder.UseShieldPotion.read_from(stream)
        if tag == DropShieldPotions.TAG:
            return ActionOrder.DropShieldPotions.read_from(stream)
        if tag == DropWeapon.TAG:
            return ActionOrder.DropWeapon.read_from(stream)
        if tag == DropAmmo.TAG:
            return ActionOrder.DropAmmo.read_from(stream)
        if tag == Aim.TAG:
            return ActionOrder.Aim.read_from(stream)
        raise Exception("Unexpected tag value")

class Pickup(ActionOrder):
    """Pick up loot"""

    TAG = 0

    __slots__ = ("loot",)

    loot: int

    def __init__(self, loot: int):
        self.loot = loot
        """Loot id"""

    @staticmethod
    def read_from(stream: StreamWrapper) -> "Pickup":
        """Read Pickup from input stream
        """
        loot = stream.read_int()
        return Pickup(loot)
    
    def write_to(self, stream: StreamWrapper):
        """Write Pickup to output stream
        """
        stream.write_int(self.TAG)
        stream.write_int(self.loot)
    
    def __repr__(self):
        return "Pickup(" + \
            repr(self.loot) + \
            ")"

ActionOrder.Pickup = Pickup

class UseShieldPotion(ActionOrder):
    """Use shield potion"""

    TAG = 1

    __slots__ = ()


    def __init__(self):
        pass

    @staticmethod
    def read_from(stream: StreamWrapper) -> "UseShieldPotion":
        """Read UseShieldPotion from input stream
        """
        return UseShieldPotion()
    
    def write_to(self, stream: StreamWrapper):
        """Write UseShieldPotion to output stream
        """
        stream.write_int(self.TAG)
    
    def __repr__(self):
        return "UseShieldPotion(" + \
            ")"

ActionOrder.UseShieldPotion = UseShieldPotion

class DropShieldPotions(ActionOrder):
    """Drop shield potions on the ground"""

    TAG = 2

    __slots__ = ("amount",)

    amount: int

    def __init__(self, amount: int):
        self.amount = amount
        """Amount of potions"""

    @staticmethod
    def read_from(stream: StreamWrapper) -> "DropShieldPotions":
        """Read DropShieldPotions from input stream
        """
        amount = stream.read_int()
        return DropShieldPotions(amount)
    
    def write_to(self, stream: StreamWrapper):
        """Write DropShieldPotions to output stream
        """
        stream.write_int(self.TAG)
        stream.write_int(self.amount)
    
    def __repr__(self):
        return "DropShieldPotions(" + \
            repr(self.amount) + \
            ")"

ActionOrder.DropShieldPotions = DropShieldPotions

class DropWeapon(ActionOrder):
    """Drop current weapon"""

    TAG = 3

    __slots__ = ()


    def __init__(self):
        pass

    @staticmethod
    def read_from(stream: StreamWrapper) -> "DropWeapon":
        """Read DropWeapon from input stream
        """
        return DropWeapon()
    
    def write_to(self, stream: StreamWrapper):
        """Write DropWeapon to output stream
        """
        stream.write_int(self.TAG)
    
    def __repr__(self):
        return "DropWeapon(" + \
            ")"

ActionOrder.DropWeapon = DropWeapon

class DropAmmo(ActionOrder):
    """Drop ammo"""

    TAG = 4

    __slots__ = ("weapon_type_index","amount",)

    weapon_type_index: int
    amount: int

    def __init__(self, weapon_type_index: int, amount: int):
        self.weapon_type_index = weapon_type_index
        """Weapon type index (starting with 0)"""
        self.amount = amount
        """Amount of ammo"""

    @staticmethod
    def read_from(stream: StreamWrapper) -> "DropAmmo":
        """Read DropAmmo from input stream
        """
        weapon_type_index = stream.read_int()
        amount = stream.read_int()
        return DropAmmo(weapon_type_index, amount)
    
    def write_to(self, stream: StreamWrapper):
        """Write DropAmmo to output stream
        """
        stream.write_int(self.TAG)
        stream.write_int(self.weapon_type_index)
        stream.write_int(self.amount)
    
    def __repr__(self):
        return "DropAmmo(" + \
            repr(self.weapon_type_index) + \
            ", " + \
            repr(self.amount) + \
            ")"

ActionOrder.DropAmmo = DropAmmo

class Aim(ActionOrder):
    """Start/continue aiming"""

    TAG = 5

    __slots__ = ("shoot",)

    shoot: bool

    def __init__(self, shoot: bool):
        self.shoot = shoot
        """Shoot (only possible in full aim)"""

    @staticmethod
    def read_from(stream: StreamWrapper) -> "Aim":
        """Read Aim from input stream
        """
        shoot = stream.read_bool()
        return Aim(shoot)
    
    def write_to(self, stream: StreamWrapper):
        """Write Aim to output stream
        """
        stream.write_int(self.TAG)
        stream.write_bool(self.shoot)
    
    def __repr__(self):
        return "Aim(" + \
            repr(self.shoot) + \
            ")"

ActionOrder.Aim = Aim